rclpy/rclpy/impl/__init__.py

```diff
@@ -17,8 +17,20 @@
 
 
 def _import(name):
+    dll_dir_handles = []
     try:
-        return importlib.import_module(name, package='rclpy')
+        # New in Python 3.8: on Windows we should call 'add_dll_directory()' for directories
+        # containing DLLs we depend on.
+        # https://docs.python.org/3/whatsnew/3.8.html#bpo-36085-whatsnew
+        if os.name == 'nt':
+            path_env = os.environ['PATH'].split(';')
```

Comment:
There is a reason why Python chose to *not* use all paths in `PATH`. Maybe it would be better to not bypass this mechanism entirely but select specific paths only to be added, e.g. `rclpy`, message packages.

Reply:
Message packages should probably contain their own logic to ensure their path is added too.

Reply:
> There is a reason why Python chose to not use all paths in PATH

What I gather from the issue[1] is that *where* DLLs are loaded from was not consistent before Python 3.8 (e.g. in general, you couldn't assume that PATH was ever searched for DLLs; it was using the process default ordering[2]), which has some security implications[3]. The change in Python 3.8 forces a consistent search order.

The reason to not include `PATH` in the search order is a little above my head, but I think it boils down to a minimizing the risk of a so-called *DLL preloading attack*. I agree we could try could try down-selecting to just what we need, though I'm not sure of a proper way to achieve this. Maybe using CMake variables that are exported by packages producing DLLs? Also, I guess there will be other DLLs installed on the system we rely on that should also be added (e.g. OpenSSL).

---

[1] https://bugs.python.org/issue36085
[2] https://docs.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order
[3] https://docs.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-security

Reply:
Yeah, you're right. `rosidl_generator_py` needs a similar patch. Pretty much anywhere we use `importlib.import_module`.